sstables/sstable_set.cc

```diff
@@ -271,7 +289,10 @@ void partitioned_sstable_set::insert(shared_sstable sst) {
     _all->insert(sst);
     auto undo_all_insert = defer([&] () { _all->erase(sst); });
 
-    _all_runs[sst->run_identifier()].insert(sst);
+    // If sstable doesn't satisfy disjoint invariant, then place it in a new sstable run.
+    while (!_all_runs[sst->run_identifier()].insert(sst)) {
+        sst->set_new_run_identifier();
```

Comment:
There's something I don't understand in the goal of this patch. We don't expect this case (bad run_identifier) to happen normally, in LCS or ICS which rely on runs working correctly. If it ever happens in LCS or ICS, it will break them, no? Or is it harmless?

Reply:
I don't think this will happen during LCS or ICS regular activity, unless there's some unknown bug which causes a run to contain overlapping files. I even gave an example, in the commit description, on this could happen when switching strategies. And this is exactly the goal of this patch. In parallel to this patch, I am working on a reader for sstable_run that should be built on the assumption that the run contains disjoint files only. By enforcing the disjointness as an invariant, then I can trust sstable_run to store disjoint files only.